### ISSUE TITLE
Investigation into the nature of JavaScript

### ISSUE THREAD

**Issue by noobling** [noobling/david-learns#2]
## JavaScript is single threaded programming language
A thunk is a function that has some closured state to keep track of value(s). Becomes a container around some state and you can pass it around anywhere in your program.

A synchronous thunk is a function that doesn't require any parameters except the callback so you can get the value out.

With an asynchronous Thunk we reduce the complexity of managing time in our program. The program state becomes independent of time.

``` JavaScript
function addAsync(x, y, cb) {
  setTimeout(function () {
    cb(x + y);
  }, 1000);
}

var asyncThunk = function(cb) {
 addAsync(10, 15, cb);
}

asyncThunk(function () {
  sum; // 25
});
```

## Blocking (synchronous) vs Non-blocking (asynchronous)

* Blocking
  * Easy to write
  * Uses multi-threading
  * Memory and context-switching overhead

* Non-blocking/event-loop
  * Single-threaded (Which means you cannot block the thread)
  * High-concurrency with low-memory consumption
  * Great for UI and IO-bound services rather than CPU-bound (Because anything using the CPU will block everything that is happening)

All modern JavaScript engines use event driven non-blocking 

* Examples of blocking calls in JS
  * Alert, prompts
  * Synchronous file reads and friends in node

### How to prevent blocking of single thread

Use callbacks. Pass a function when the task is complete

## Promises

They are a thin layer of abstraction over callbacks. They help with readability because you don't have that sort of deep nesting if you need to do things in sequence when using callbacks.

``` JavaScript
sleep(1000)
  .then(() => { console.log('one'); return sleep(1000);}) 
  .then(() => { console.log('two'); return sleep(1000);})
  .then(() => { console.log('three'); })
```
By returning a promise we can call then on it.

Here is another example of the power of promises

``` Javascript
fetchJSON('/user-profile')
  .then((user) => {
    return fetchJSON(`/users/${user.id}/friends`);
  })
  .then((friendIDs) => { 
    let promises = friendIDs.map((id) => {
      return fetchJSON(`/users/${id}`)
    });
    return Promise.all(promises);
  })
  .then((friends) => console.log(friends));
```
Here we have to do somethings sequentially and other things parallel. For example we wait until we get the user id in order to get the friends associated with it. We can then retrieve the profile of each friend in parallel by iterating over all the friends. `Promise.all(iterable)` returns a single promise when all the promises have inside the *iterable* have been resolved.

## Await

We can achieve synchronous looking code with `await`. This has been achieved through `generator functions`. This idea is not related to aysnc or event loops and not about concurrency. They are just about pausing the function.

``` JavaScript
function * generatorFunc() {
  let result = fetch('/users')
  // Pause execution by yielding.
  yield result;
  // Later something caused us to resume.
  console.log('We are back');
}
```

``` JavaScript
async function getUsers() { 
  let result = await fetchJSON('/users') // This will pause the whole function
  console.log(result);
}
```

A `promise` always has to follow `await`. Can only use await inside a async function.
Async function itself returns a promise. When we await a promise the function pauses until the promise is ready (resolved)

``` JavaScript
async function getUserFriends() {
  let user = await fetchJSON('/users/user-profile')
  let friendsID = await fetchJSON(`/users/${user.id}/friends`)
  let promises = friendsID.map((id) => { return fetchJSON(`/users/${id}`) })
  let friends = await Promise.all(promises)
  console.log(friends)
}

let friends = await getUserFriends()
doSomethingCallWithFriends(friends)
```
No need for callbacks now with `aysnc/await`

## Event Loop

Event loop is what allows Node.js to perform non-blocking I/O operations -- despite the fact that JavaScript is single-threaded -- by off loading operations to the system kernel whenever possible.

Since most modern kernels are multi-threaded, they can handle multiple operations in the background.

### Event loop operations
<pre>
<code>
   ┌───────────────────────────┐
┌─&gt;│           timers          │
│  └─────────────┬─────────────┘
│  ┌─────────────┴─────────────┐
│  │     pending callbacks     │
│  └─────────────┬─────────────┘
│  ┌─────────────┴─────────────┐
│  │       idle, prepare       │
│  └─────────────┬─────────────┘      ┌───────────────┐
│  ┌─────────────┴─────────────┐      │   incoming:   │
│  │           poll            │&lt;─────┤  connections, │
│  └─────────────┬─────────────┘      │   data, etc.  │
│  ┌─────────────┴─────────────┐      └───────────────┘
│  │           check           │
│  └─────────────┬─────────────┘
│  ┌─────────────┴─────────────┐
└──┤      close callbacks      │
   └───────────────────────────┘
</pre>
</code>

Each phase has a FIFO queue of callbacks to execute

#### Phases Overview

* **Times** This phase executes callbacks scheduled by `setTimeout()` and `setInterval()`
* **pending callbacks** executes I/O callbacks deferred to next loop iteration
* **idle, prepare** only used internally
* **poll** Retrieve new I/O events; execute I/O related callbacks (These are most callbacks with the exception of close callbacks and the ones scheduled by timers and `setImmediate()`
* **check** `setImmediate()` callbacks invoked here
* **close callbacks** some close callbacks `socket.on('close')`

#### Timers

They run after a delay specified by the user however this it is not guaranteed to execute exactly after that time because other callbacks and OS scheduling may delay it.

### Closure

A way to remember and continue to access a function's scope (it's variables) even once the function has finished running.

Most common usage of closure in JavaScript is the module pattern. They let you define private implementation details (variables, functions) as well as a public API that is accessible from the outside world

### this

`this` normally points to an object but which object it points to depends on how the function was called

## How is JavaScript Asynchronous and single-threaded?

JavaScript language is single-threaded and the asynchronous behaviour is not part of the JavaScript language itself rather they are built on top of the core JavaScript language in the browser and accessed through the browser APIs.

``` JavaScript
function main(){
  console.log('A');
  setTimeout(
    function display(){ console.log('B'); }
  ,0);
  console.log('C');
}
main();
```

The code will be executed like this underneath the hood, that is why 'B' is actually displayed last

![](/noobling/david-learns/blob/master/js-event-loop.png?raw=true)

JavaScript is single threaded but the browser offers web apis that can execute code asynchronously for you. In node this would be C++ library calls. `setTimeout` is one such api call offered to us.

The event loop is very simple it just looks at the task queue and it looks at the stack if the stack is empty it pushes the first thing in the task queue onto the stack.

## Weird shit

* Return needs to have something on the same line with it otherwise nothing is returned
``` js
return (
  <div> Hello world </div>
)
```
Won't work
``` js
return 
(
  <div> Hello world </div>
)
```
## IIFE (immediately invoked function expression) JS pattern
Allows us to create scope in various places and doesn't pollute the namespace
```js
var foo = 'foo1'
(function notInGlobalNameSpace() {
  var foo = 'foo2'
  console.log('did something')
}) ()
console.log(foo)
```

## Compiler

JavaScript compiles the code on the fly the runs them. 

The general steps to compiling JS program are as follows

1. **Tokenizing/Lexing** breaking up a string of characters into meaningful chunks called tokens. 
`var a = 2;` -> `var` `a` `=` `2` `;`
2. **Parsing** Taking a stream (array) of tkens and turning it into a tree of nested elements which collectively represent the grammatical structure of the program. This tree is called an AST (Abstract Syntax Tree)
3. **Code-Generation** This is where memory allocation occurs and how your code turns into machine code to be executed [example](https://github.com/jamiebuilds/the-super-tiny-compiler)

## Low-level implementation of how scoping occurs

1. Engine: responsible for start to finish the compilation and execution of our JS program
2. Compiler: one of Engine's friends: handles all the dirty work for parsing and code-generation
3. Scope: another friend of Engine; collects and maintains a look-up list of all the declared identifiers (variables) and enforces rules as to how these are accessible to currently executing

## Sources
https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/ (Event loop in node js)
https://frontendmasters.com/courses/rethinking-async-js/ (Thunks)
https://www.youtube.com/watch?v=NsQ2QIrQShU (Callbacks, async/await, promises)
https://www.youtube.com/watch?v=8aGhZQkoFbQ (Event loops explained)
